Source/WebCore/editing/cocoa/HTMLConverter.mm

```diff
@@ -2451,15 +2451,21 @@ static bool hasAncestorQualifyingForWritingToolsPreservation(Element* ancestor,
     if (!ancestor)
         return false;
 
-    return cache.ensure(*ancestor, [&] {
-        if (isMailBlockquote(*ancestor))
-            return true;
+    if (!cache.contains(*ancestor)) {
+        auto result = [&] {
+            if (isMailBlockquote(*ancestor))
+                return true;
 
-        if (auto renderer = ancestor->renderer(); renderer && renderer->style().whiteSpace() == WhiteSpace::Pre)
-            return true;
+            if (auto renderer = ancestor->renderer(); renderer && renderer->style().whiteSpace() == WhiteSpace::Pre)
+                return true;
+
+            return hasAncestorQualifyingForWritingToolsPreservation(ancestor->parentElement(), cache);
+        }();
+
+        cache.set(*ancestor, result);
+    }
 
-        return hasAncestorQualifyingForWritingToolsPreservation(ancestor->parentElement(), cache);
-    }).iterator->value;
+    return cache.get(*ancestor);
```

Comment:
This will do the two hash map lookups and bloats the code. Can we use `add` instead (and check isNewEntry)?

Reply:
@rniwa I made the change here, but had a question.

Since we'd be calling `add` recursively, and then writing to `addResult.iterator->value`, couldn't potential rehashing make the result invalid?

Reply:
As we discussed on slack, this wouldn't work due to the order of operations required here.